src/Http/Transport/CurlTransport.php

```diff
@@ -0,0 +1,196 @@
+        CURLOPT_TIMEOUT        => self::DEFAULT_TIMEOUT,
+    ];
+
+    /**
+     * These options must be always set to the given value to ensure that the transport operates properly.
+     */
+    private const INTERNAL_CURL_OPTIONS = [
+        CURLOPT_ENCODING       => 'UTF-8',
+        CURLOPT_RETURNTRANSFER => 1,
+        CURLOPT_FOLLOWLOCATION => 1,
+    ];
+
+    /**
+     * Time to connect to BigBlueButton host.
+     */
+    private const DEFAULT_CONNECT_TIMEOUT = 10;
+
+    /**
+     * Time to for the whole request and response flow (inclusive connect, see above).
+     */
+    private const DEFAULT_TIMEOUT = 30;
+
+    /**
+     * @var mixed[]
+     */
+    private $curlOptions;
+
+    /**
+     * Allows to inject custom cURL options used on dispatching request to BBB.
+     * Please note that you must ensure on your own that the usage of custom options does not break the transport.
+     *
+     * @see CurlTransport::createWithDefaultOptions() Please use this factory method to create the CurlTransport with the recommended defaults.
+     *
+     * @link https://www.php.net/manual/en/function.curl-setopt.php List of known cURL options
+     * @param mixed[] $curlOptions A list of cURL options to pass to the cURL handle. Option name as key, option value as value.
+     */
+    public function __construct(array $curlOptions = [])
+    {
+        $this->curlOptions = $curlOptions;
+    }
+
+    /**
+     * @param  string[] $additionalCurlOptions A list of additional cURL options to pass to the cURL handle. Option name as key, option value as value.
+     * @return self
+     */
+    public static function createWithDefaultOptions(array $additionalCurlOptions = []): self
+    {
+        // @codeCoverageIgnoreStart
+        return new self(ArrayHelper::mergeRecursive(true, $additionalCurlOptions, self::DEFAULT_CURL_OPTIONS));
+        // @codeCoverageIgnoreEnd
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function request(TransportRequest $request): TransportResponse
+    {
+        // @codeCoverageIgnoreStart
+        if (!extension_loaded('curl')) {
+            throw new RuntimeException('Curl PHP module is not installed or not enabled.');
+        }
+        // @codeCoverageIgnoreEnd
+
+        $ch = curl_init();
+        // @codeCoverageIgnoreStart
+        if (!$ch) {
+            throw new RuntimeException('Could not create curl instance. Error: ' . curl_error($ch));
+        }
+        // @codeCoverageIgnoreEnd
+
+        // Needed to store the JSESSIONID
+        $cookieFile = @tmpfile(); // Silenced as tmpfile can throw notices like "creating file in system temporary directory".
+        // @codeCoverageIgnoreStart
+        if (false === $cookieFile) {
+            throw new RuntimeException('Could not create temporary file for cookies.');
+        }
+        // @codeCoverageIgnoreEnd
+        $cookieFilePath = stream_get_meta_data($cookieFile)['uri'];
+
+        $options = ArrayHelper::mergeRecursive(
```

Comment:
If you set the http header for Content-type or Content-length as shown in this example, you get duplicates

```php
$request   = new TransportRequest('http://localhost:8057/dump.php', 'FOO', 'application/xml');
$transport = new CurlTransport([CURLOPT_HTTPHEADER => ['X-Foo: Bar', 'X-Bar: Foo','Content-length: 5']]);
$response = $transport->request($request);
```


```php
array(5) {
    [0] =>
    string(29) "Content-type: application/xml"
    [1] =>
    string(17) "Content-length: 3"
    [2] =>
    string(10) "X-Foo: Bar"
    [3] =>
    string(10) "X-Bar: Foo"
    [4] =>
    string(17) "Content-length: 5"
  }
```

Reply:
A solution could be to parse all headers to an array, merge/overwrite and reassemble.

Reply:
Good catch! I guess it would be fine to implement that.

Reply:
I added a specialized header util  for that case plus a test of the scenario you have described.